### ISSUE TITLE
Determining the communicator from an MPI_Message object.

### ISSUE THREAD

**Issue by SpinTensor** [mpi-forum/mpi-issues#323]
# Problem

Currently there is no way to determine the used communicator from an `MPI_Message` object.
If an `MPI_Mrecv` is called the status variable is able to provide the source as `status.MPI_SOURCE`, but it is impossible to translate this rank to the rank in e.g. `MPI_COMM_WORLD`. The communicator is only given to `MPI_Mprobe`.
This is especially bad for profiling layers, where inside the `MPI_Mprobe` function there is definitively no access to the communicator from the probe.

# Proposal

I propose a function like `int MPI_Get_comm_from_message(MPI_Comm *comm, MPI_Message message)` or something nicer sounding. 
`message` is the input and after the function returns comm contains the communicator handle in which the message was probed.

# Changes to the Text

The addition to the text would be the precise definition of the above mentioned function.

# Impact on Implementations

Implementations like OpenMPI already store a pointer to the communicator in the opaque MPI_Message object.
Thus, it would be easy to implement a function that simply returns a copy of that pointer.

# Impact on Users

Users would be able to use additional information in the context of `MPI_Mrecv` e.g. to determine the global (or in other communicators) rank of the source rank.

# References

See MPI-4.0 draft, which does not contain such functionality.

**Comment by jeffhammond** [comment]
This isn't a bad idea.  @wrwilliams or @mahermanns might be interested from the profiling perspective.  One can argue in the application, that the use case we expect for Mprobe+Mrecv allows the user to know the Comm and the Message easily enough, but I agree that it's not feasible in the context of profiling.

**Comment by wrwilliams** [comment]
As I see it, there's precisely one use case that this helps:

* User wants to intercept `MRecv`
* User needs communicator info that would otherwise be collected and associated with the message in an `MProbe` wrapper
* User wants to avoid intercepting `MProbe` to reduce overhead(?)

It pretty well parallels the extra work a tool has to do with non-blocking request tracking, IMO--which is not to say that such work shouldn't be avoided when the implementation already knows the answer trivially. I know that in Score-P we currently do this work for non-blocking P2P/collectives but do not do it for MProbe/MRecv; I've asked for input from the team as to whether this would help any users.

**Comment by jeffhammond** [comment]
it's not just that they don't want to intercept MProbe but that profiling tools really don't have the ability to have state persist between calls.  the whole point of the M*** APIs is for thread-safety, nor can we assume that calls to Mprobe and Mrecv occur in close temporal proximity.  that's certainly the common case, but i'm allowed to be a psychopath and call Mprobe a million times and then call Mrecv a million times after randomize the access order of my Message array.

the counterargument to this feature is that rank translation isn't a good motivation for having access to this information.  i don't have a position on that right now.  i merely support the idea that there is no reasonable way to associate a Comm with a Message in the important use case of profiling.

**Comment by mahermanns** [comment]
We have been discussing this in the tools working group.

First, currently tools to have state persist between calls (otherwise we couldn't match an `MPI_Isend` and its completing `MPI_Wait`.

Picking up @wrwilliams 's point of saving overhead, the problematic call would rather be `MPI_Improbe` that has the potential to generate a lot of overhead with unsuccessful invocations. If tools could ignore wrapping the `MPI_Improbe` altogether and then get all the critical information (i.e., the message envelope) from the message object, it would actually help to reduce overhead and we don't have to track message objects at all (i.e., getting rid of a hash-table insert and lookup).

Thus, we would actually go a step further and propose something along the lines of:
```c
int MPI_Message_get_envelope(MPI_Message message, int *rank, int *tag, MPI_Comm *comm);
```

Implementations should be able to return this information with low overhead, as it should be readily available for `MPI_Mrecv` and the like.

**Comment by mahermanns** [comment]
As per discussion in the working group:

The `rank` and `tag` value that were actually matched can be obtained from the status object. Also that information needs to be still available during the `MPI_Mrecv`, as the MPI library needs to fill the status object.

If access to the arguments of the actual `MPI_Improbe` is needed, a tool would still have to wrap `MPI_Improbe`. Situations where this is still needed might include:
- A tracing tool would like to record wildcard arguments provided to calls (to be used later by a consumer of the trace)
- A correctness tool may want to query the `status` before the actual `MPI_Mrecv` and warn about possible overruns
- A tool may want to differentiate the `ERROR` part of the different status objects returned by the probe vs. the recv.

Apart for the scenarios above (or similar) the communicator would suffice to obtain for a performance tool that wants to reduce overhead. Resulting in the following function call:

```c
int MPI_Message_get_comm(MPI_Message message, MPI_Comm *comm);
```

The group leans to have this smaller call. Potential use cases for other information regarding the message object could and should still be gathered, but we'd move forward with this proposal now.

**Comment by mahermanns** [comment]
### From the discussion in the Tools WG call on 2024-07-08:
- Maybe returning the full envelope is the cleaner API call
  - What about returning potential wildcards
- Should we make the communicator a part of the `MPI_Status`
   - This is more intrusive
   - But then all envelope information would be in one place 
   - What about the error code?
      - The message object is only created from envelope information (no buffer handling etc.)
      - Not much to do with this information. 
- What about when the communicator is already freed (by the user) when the message is received?
  - Is `MPI_COMM_NULL` returned?
    - Is that useful for tools then? (usually no, because the communicator handle is used as key for information lookup)
  - If not, is the communicator usable for new communication?
    - if it's stale, how to tell the user **not** to use it (would be precedence for MPI handing back something stale)
      - Any copies of handles are also stale, but there it's the users task to take care of their proper use
      - Can this be addressed sufficiently with an *Advice to Users*
      - Can this be handled by a specific well-defined error code and behavior? Is there something special to keep in mind from the fault-tolerance group?
      - Does the user have to free the communicator handle returned?
        - Do we need an `MPI_Comm_is_freed`? ... How would one implement this? The memory may already be freed (or implementations will no longer be able to free the memory behind a freed communicator)
- Is this call in the right chapter?
  - There is precedence for some status information to be placed in the "external interfaces" chapter, for people who are interested in using generalized requests
  - Could be part of the (a) p2p, (b) external interfaces, (c) or tools chapter. Which location is best?

**Comment by jeffhammond** [comment]
> Should we make the communicator a part of the MPI_Status

No. ABI impact is unpleasant. It's not something most uses of status require.

**Comment by bosilca** [comment]
I can see how this would be interesting to have for tools developers, but for anyone else this is a terrible, and unnecessary idea.

First and foremost because it  provides the user with another reference to the communicator, and it is not clear how or for how long the user can hold onto it. Currently the life cycle of a communicator is between a creation function and the corresponding free, allowing correct applications to retrieve and release a communicator only once. Yes, operations might be pending, but nothing that allows the user to regain access to the communicator itself. Thus, once all pending operations are gone the communicator can be safely discarded by the implementation.

With this new function, it looks as each call could (or should) return a new reference and possibly increase the refcount of the communicator. An advice to user is not an encompassing solution, and assuming all handles become stale is horribly unsafe because we are back at either accepting segfaults or having full support for unique pointers (this has been discussed in another issue, and we are still waiting for a plausible implementation).

**Comment by mahermanns** [comment]
@bosilca I can see how this can be a problem. Because we return a handle, we'd have to specify the life time and ownership of the handle (i.e., who has to free the handle) :thinking: .... I wonder if it were possible to specify that the call behaves as if the user had copied the handle to another variable.

```c
MPI_Comm commA, commB;
MPI_Comm_dup(MPI_COMM_WORLD, &commA);
commB = commA;
MPI_Comm_free(&commA); // commB is also no longer valid
```
AFAIK, this code is valid (yet creates an aliasing that may be considered a code smell :wink:) ... We will need to discuss this a bit further and see if there are solutions to your concerns (which I think are all valid) ... maybe there is a different solution to the initial problem.

**Comment by bosilca** [comment]
There is no aliasing in MPI. Indeed, a dup of a communicator is not just a refcount, it has to enable a different message matching. That being said MPI does not have a concept of refcount exposed to the user, even if most of the implementations need one internally (to know when pending operations have released internal resources).

**Comment by jprotze** [comment]
I would bind the life-time of this comm handle to the life-time of the message.

At the same time, I think this function puts unnecessary burden on the MPI implementation and constrains implementation choices.

**Comment by devreal** [comment]
We could provide IDs that identify communicators. We have those internally already. They may or may not be unique throughout the life-time of the application and are scoped on the process (not globally) but that should not be hard to handle in tools. The message could expose the communicator ID that tools can match with an internal list of communicators they got through something like `MPI_Comm_get_id(&id)`

**Comment by jeffhammond** [comment]
> I would bind the life-time of this comm handle to the life-time of the message.

The communicator object must still exist as long as the message does or else the message cannot be received.

The handle doesn't matter because the query procedure returns the handle value to user code.

**Comment by mahermanns** [comment]
#### Discussion of the MPI tools WG on Jul 22, 2024
- Overhead through interception of `MPI_Improbe` should be small enough for practical use-cases if only a timestamp is taken at the beginning of the wrapper and then all remaining tool code is only executed on successful return from `MPI_Improbe`.
- This call would imply that the MPI library would **have to** store the communicator in the message object and delay destruction through additional ref-counting
- Lifetime should (if at all) be limited to the lifetime of the corresponding `MPI_Message` object (until right before the call to `PMPI_Mrecv`
- Function would give precedence to "returning a reference to a comm handle to the user" that the user would not have to free (all other handles returned from MPI function have to be freed by the user)
- Consensus in the group is therefore to abandon this until new good arguments are presented for this call
